File: src/main/java/sawyern/cookiebot/services/CookieService.java
package sawyern.cookiebot.services;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import sawyern.cookiebot.constants.CookieType;
import sawyern.cookiebot.models.dto.GiveCookieDto;
import sawyern.cookiebot.models.entity.Account;
import sawyern.cookiebot.models.entity.Cookie;
import sawyern.cookiebot.exception.CookieException;
import sawyern.cookiebot.repository.CookieRepository;

import java.util.Collection;
import java.util.stream.Collectors;

@Service
@Slf4j
@RequiredArgsConstructor
public class CookieService {

    private final CookieRepository cookieRepository;
    private final AccountService accountService;

    public void giveCookieTo(GiveCookieDto giveCookieDto) throws CookieException {
        if (giveCookieDto.getNumCookies() <= 0)
            throw new CookieException("Invalid number of cookies", HttpStatus.BAD_REQUEST);
        int cookies = getAllCookiesForAccount(giveCookieDto.getSenderId());
        if (cookies < giveCookieDto.getNumCookies())
            throw new CookieException("Not enough cookies to give", HttpStatus.BAD_REQUEST);

        for (int i = 0; i < giveCookieDto.getNumCookies(); i++) {
            removeCookieOfType(giveCookieDto.getSenderId(), CookieType.NORMAL);
            generateCookie(giveCookieDto.getRecieverId(), CookieType.NORMAL);
        }
    }

    public int getAllCookiesForAccount(String discordId) {
        Collection<Cookie> cookies = cookieRepository.findByAccountDiscordId(discordId);
        return cookies.size();
    }

    public void removeCookieOfType(String discordId, String type) throws CookieException {
        Collection<Cookie> cookies = cookieRepository.findByAccountDiscordId(discordId);
        cookies = cookies.stream()
                .filter(cookie -> cookie.getType().equals(type))
                .collect(Collectors.toList());

        if (cookies.isEmpty())
            throw new CookieException("No cookies to remove.");

        try {
            cookieRepository.delete(cookies.stream().findFirst().orElseThrow(() -> new CookieException("Cookie not found")));
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            throw new CookieException("Failed to delete cookie");
        }
    }

    public void generateCookie(String discordId, String type, int num) throws CookieException {
        for (int i = 0; i < num; i++) {
            generateCookie(discordId, type);
        }
    }

    public Cookie generateCookie(String discordId) throws CookieException {
        return generateCookie(discordId, CookieType.NORMAL);
    }

    public Cookie generateCookie(String discordId, String type) throws CookieException {
        Account account = accountService.getAccount(discordId);
        Cookie cookie = new Cookie(account, type);
        try {
            cookieRepository.save(cookie);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            throw new CookieException("Error saving cookie");
        }
        return cookie;
    }
}

File: src/main/kotlin/sawyern/cookiebot/services/CookieService.kt
package sawyern.cookiebot.services

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Lazy
import org.springframework.stereotype.Service
import sawyern.cookiebot.exception.CookieException
import sawyern.cookiebot.models.Cookie
import sawyern.cookiebot.models.Season
import sawyern.cookiebot.repository.CookieRepository

@Service
class CookieService @Autowired constructor(
        private val cookieRepository: CookieRepository,
        private val seasonService: SeasonService,
        @Lazy
        private val accountService: AccountService

) {
    private val logger: Logger = LoggerFactory.getLogger(CookieService::class.java)
    
    fun giveCookieTo(senderId: String, receiverId: String, numCookies: Int) {
        if (numCookies <= 0)
            throw CookieException("Invalid number of cookies: $numCookies")

        val senderCookies = cookieRepository.findByAccountDiscordIdAndSeason(senderId, seasonService.getCurrentSeason())

        if (senderCookies.size < numCookies)
            throw CookieException("Not enough cookies to give")

        var cookiesGiven = 0
        val receiverAccount = accountService.getAccount(receiverId)
        for (cookie in senderCookies) {
            cookie.account = receiverAccount
            cookieRepository.save(cookie)
            cookiesGiven++

            if (cookiesGiven == numCookies)
                break
        }

        if (logger.isInfoEnabled) {
            val senderAccount = accountService.getAccount(senderId)
            logger.info("Transferred $numCookies cookies from ${senderAccount.username} to ${receiverAccount.username}")
        }
    }

    fun getCookiesForAccount(discordId: String, season: Season = seasonService.getCurrentSeason()): Int {
        return cookieRepository.findByAccountDiscordIdAndSeason(discordId, season).size
    }

    fun generateCookie(discordId: String, numCookies: Int = 1, season: Season = seasonService.getCurrentSeason()) {
        val account = accountService.getAccount(discordId)

        for(i in 1..numCookies) {
            val cookie = Cookie(account, season)
            cookieRepository.save(cookie)
        }

        logger.info("Generated $numCookies cookies for ${account.username}")
    }
}